operator/pkg/ingress/ingress_class.go

```diff
@@ -4,172 +4,199 @@
+// processEvents pulls events from the IngressClasses resource and passes them to the appropriate
+// handles.
+// It returns when the given context is cancelled.
+func (i *ingressClassManager) processEvents(ctx context.Context) error {
 	var err error
-	switch ev := event.(type) {
-	case ingressClassAddedEvent:
-		log.WithField(logfields.IngressClass, ev.ingressClass.Name).Debug("Handling ingress class added event")
-		err = i.handleIngressClassAddedEvent(ev)
-	case ingressClassUpdatedEvent:
-		log.WithField(logfields.IngressClass, ev.newIngressClass.Name).Debug("Handling ingress class updated event")
-		err = i.handleIngressClassUpdatedEvent(ev)
-	case ingressClassDeletedEvent:
-		log.WithField(logfields.IngressClass, ev.ingressClass.Name).Debug("Handling ingress class deleted event")
-		err = i.handleIngressClassDeletedEvent(ev)
-	default:
-		err = fmt.Errorf("received an unknown event: %s", ev)
+
+	for {
+		select {
+		case event := <-i.ingressClassEvents:
+			err = nil
+
+			switch event.Kind {
+			case resource.Sync:
+				err = i.handleSyncEvent()
+			case resource.Upsert:
+				err = i.handleUpsertEvent(event)
+			case resource.Delete:
+				err = i.handleDeleteEvent(event)
+			}
+
+			if event.Done != nil {
+				event.Done(err)
+			}
+		case <-ctx.Done():
+			return ctx.Err()
+		}
 	}
-	return err
 }
 
-func (i *ingressClassManager) handleIngressClassAddedEvent(event ingressClassAddedEvent) error {
-	log.WithField(logfields.IngressClass, event.ingressClass).Debug("Handling ingress class add")
-	i.notify(event.ingressClass)
-	return nil
-}
+// handleSyncEvent handles a Sync event for the events of an IngressClasses Resource.
+func (i *ingressClassManager) handleSyncEvent() error {
+	log.Debug("Handling IngressClass Sync event")
+	i.synced.Store(true)
 
-func (i *ingressClassManager) handleIngressClassUpdatedEvent(event ingressClassUpdatedEvent) error {
-	log.WithField("old", event.oldIngressClass).WithField("new", event.newIngressClass).Debug("Handling ingress class update")
-	i.notify(event.newIngressClass)
 	return nil
 }
 
-func (i *ingressClassManager) handleIngressClassDeletedEvent(event ingressClassDeletedEvent) error {
-	log.WithField(logfields.IngressClass, event.ingressClass).Debug("Handling ingress class delete")
+// handleUpsertEvent handles an Upsert event for an IngressClass resource.
+// If the IngressClass has a name that is not the Cilium IngressClass name, it will be ignored.
+// If an error is returned, then the IngressClass contained in the event should be retried at a later
+// point in time.
+func (i *ingressClassManager) handleUpsertEvent(event resource.Event[*slim_networkingv1.IngressClass]) error {
+	log.WithField(logfields.IngressClass, event.Object).Warn("Handling IngressClass Upsert event")
 
-	if event.ingressClass.GetName() == ciliumIngressClassName {
-		i.ingressQueue.Add(ciliumIngressClassDeletedEvent(event))
+	if event.Object == nil || (event.Object != nil && !isIngressClassCilium(event.Object)) {
```

Comment:
no need to check `event.Object` being not nil in the second part of the condition.

-> `if event.Object == nil || !isIngressClassCilium(event.Object) {`

Reply:
Ah yeah good point.